Mathlib/RingTheory/Flat/FaithfullyFlat.lean

```diff
@@ -163,6 +169,189 @@ lemma iff_flat_and_lTensor_reflects_triviality :
+      · rintro ⟨x, rfl⟩
+        induction x using TensorProduct.induction_on <;> aesop
+      · rw [LinearMap.exact_iff.1 ex]
+        rintro ⟨x, rfl⟩
+        induction x using TensorProduct.induction_on with
+        | zero => simp
+        | tmul a b =>
+          exact ⟨⟨l12 a, complex (by simp)⟩ ⊗ₜ b, rfl⟩
+        | add x y hx hy => rw [map_add]; exact Submodule.add_mem _ hx hy,
+    LinearMap.rTensor, TensorProduct.map_range_eq_span_tmul, Submodule.span_le]
+  rintro _ ⟨x, y, rfl⟩
+  exact ⟨⟨l12 x, complex (by simp)⟩ ⊗ₜ y,
+    ⟨⟨⟨⟨l12 x, complex (by simp)⟩, ⟨⟨l12 x, by simp⟩, rfl⟩⟩ ⊗ₜ y, rfl⟩, rfl⟩⟩
+
+lemma lTensor_reflects_exact [fl : FaithfullyFlat R M]
+    (N1 N2 N3 : Type*)
+    [AddCommGroup N1] [Module R N1]
+    [AddCommGroup N2] [Module R N2]
+    [AddCommGroup N3] [Module R N3]
+    (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3)
+    (ex : Function.Exact (l12.lTensor M) (l23.lTensor M)) :
+    Function.Exact l12 l23 :=
+  rTensor_reflects_exact R M _ _ <| ex.of_ladder_linearEquiv_of_exact
+    (e₁ := TensorProduct.comm _ _ _) (e₂ := TensorProduct.comm _ _ _)
+    (e₃ := TensorProduct.comm _ _ _) (by ext; rfl) (by ext; rfl)
+
+end arbitrary_universe
+
+lemma implies_iff_exact [fl : FaithfullyFlat R M]
+    (N1 N2 N3 : Type max u v)
+    [AddCommGroup N1] [Module R N1]
+    [AddCommGroup N2] [Module R N2]
+    [AddCommGroup N3] [Module R N3]
+    (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3) :
+    Function.Exact l12 l23 ↔ Function.Exact (l12.rTensor M) (l23.rTensor M) :=
+  ⟨fun e => Module.Flat.iff_rTensor_exact.1 fl.toFlat e,
+    fun ex => rTensor_reflects_exact R M l12 l23 ex⟩
+
+lemma iff_iff_rTensor_exact :
+    FaithfullyFlat R M ↔
+    (∀ (N1 N2 N3 : Type max u v)
+        [AddCommGroup N1] [Module R N1]
+        [AddCommGroup N2] [Module R N2]
+        [AddCommGroup N3] [Module R N3]
+        (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3),
+        Function.Exact l12 l23 ↔ Function.Exact (l12.rTensor M) (l23.rTensor M)) :=
+  ⟨fun fl => implies_iff_exact R M, fun iff_exact =>
+    iff_flat_and_rTensor_reflects_triviality _ _ |>.2 ⟨Flat.iff_rTensor_exact.2 <| by aesop,
+    fun N _ _ h => subsingleton_iff_forall_eq 0 |>.2 <| fun y => by
+      simpa [eq_comm] using (iff_exact PUnit N PUnit 0 0 |>.2 fun x => by
+        simpa using Subsingleton.elim _ _) y⟩⟩
+
+lemma iff_iff_lTensor_exact :
+    FaithfullyFlat R M ↔
+    (∀ (N1 N2 N3 : Type max u v)
+        [AddCommGroup N1] [Module R N1]
+        [AddCommGroup N2] [Module R N2]
+        [AddCommGroup N3] [Module R N3]
+        (l12 : N1 →ₗ[R] N2) (l23 : N2 →ₗ[R] N3),
+        Function.Exact l12 l23 ↔ Function.Exact (l12.lTensor M) (l23.lTensor M)) :=
+  iff_iff_rTensor_exact _ _ |>.trans <| iff_of_eq <|
+    forall_congr <| fun N1 => forall_congr fun N2 => forall_congr fun N3 =>
+    forall_congr fun _ => forall_congr fun _ => forall_congr fun _ => forall_congr fun _ =>
+    forall_congr fun _ => forall_congr fun _ => forall_congr fun l12 => forall_congr fun l23 =>
```

Comment:
There must be a better way to do this. This looks awful.

Reply:
Maybe it can be simplified by using [forall₅_congr](https://leanprover-community.github.io/mathlib4_docs/Init/PropLemmas.html#forall%E2%82%85_congr)

Reply:
I introduced a lemma saying `exact (f \ox M) (g \ox M)` iff `exact (M \ox f) (M \ox g)`

Reply:
Yes this approach is much better!